File: sdk/swift/Sources/MeshLLM/NativeRuntime.swift
import Foundation

#if canImport(MeshLLMFFI)
import MeshLLMFFI

public typealias NativeRuntimeInstallOptions = NativeRuntimeInstallOptionsNative
public typealias NativeRuntimeInstallOutcome = NativeRuntimeInstallOutcomeNative
public typealias NativeRuntimeDownloadProgress = NativeRuntimeDownloadProgressNative
public typealias InstalledNativeRuntime = InstalledNativeRuntimeNative
public typealias NativeRuntimePruneResult = NativeRuntimePruneResultNative
public typealias NativeRuntimeVerificationPolicy = NativeRuntimeVerificationPolicyNative
public typealias NativeRuntimePruneMode = NativeRuntimePruneModeNative

public struct NativeRuntimeResolveOptions: Sendable {
    public let artifactDirectory: URL?
    public let searchDirectories: [URL]
    public let cacheDirectory: URL?
    public let allowDownload: Bool
    public let selection: String

    public init(
        artifactDirectory: URL? = nil,
        searchDirectories: [URL] = [],
        cacheDirectory: URL? = nil,
        allowDownload: Bool = false,
        selection: String = "recommended"
    ) {
        self.artifactDirectory = artifactDirectory
        self.searchDirectories = searchDirectories
        self.cacheDirectory = cacheDirectory
        self.allowDownload = allowDownload
        self.selection = selection
    }
}

public enum NativeRuntime {
    public static var meshVersion: String {
        currentMeshVersion()
    }

    public static var skippyAbiVersion: String {
        currentSkippyAbiVersion()
    }

    public static func install(
        _ options: NativeRuntimeInstallOptions,
        onProgress: (@Sendable (NativeRuntimeDownloadProgress) -> Void)? = nil
    ) async throws -> NativeRuntimeInstallOutcome {
        let progress = onProgress.map(ProgressListener.init)
        return try await runBlocking {
            try installNativeRuntime(options: options, progress: progress)
        }
    }

    public static func resolve(
        _ options: NativeRuntimeResolveOptions = NativeRuntimeResolveOptions(),
        onProgress: (@Sendable (NativeRuntimeDownloadProgress) -> Void)? = nil
    ) async throws -> InstalledNativeRuntime {
        var bundleDirs: [String] = options.searchDirectories.map(\.path)
        if let artifactDirectory = options.artifactDirectory {
            bundleDirs.insert(artifactDirectory.path, at: 0)
        }
        let outcome = try await install(
            NativeRuntimeInstallOptions(
                meshVersion: nil,
                skippyAbiVersion: nil,
                selection: options.selection,
                manifestPath: nil,
                manifestUrl: nil,
                bundleDirs: bundleDirs,
                cacheDir: options.cacheDirectory?.path,
                verificationPolicy: .requireChecksum,
                allowDownload: options.allowDownload
            ),
            onProgress: onProgress
        )
        return outcome.runtime
    }

    public static func validate(
        artifactDirectory: URL,
        cacheDirectory: URL? = nil
    ) async throws -> InstalledNativeRuntime {
        try await resolve(
            NativeRuntimeResolveOptions(
                artifactDirectory: artifactDirectory,
                cacheDirectory: cacheDirectory,
                allowDownload: false
            )
        )
    }

    public static func installed(cacheDirectory: URL? = nil) async throws -> [InstalledNativeRuntime] {
        try await runBlocking {
            try installedNativeRuntimes(cacheDir: cacheDirectory?.path)
        }
    }

    @discardableResult
    public static func remove(
        meshVersion: String,
        nativeRuntimeId: String,
        cacheDirectory: URL? = nil
    ) async throws -> Bool {
        try await runBlocking {
            try removeNativeRuntime(
                cacheDir: cacheDirectory?.path,
                meshVersion: meshVersion,
                nativeRuntimeId: nativeRuntimeId
            )
        }
    }

    public static func prune(
        cacheDirectory: URL? = nil,
        activeMeshVersion: String? = nil,
        mode: NativeRuntimePruneMode = .keepActiveAndPrevious
    ) async throws -> NativeRuntimePruneResult {
        try await runBlocking {
            try pruneNativeRuntimes(
                cacheDir: cacheDirectory?.path,
                activeMeshVersion: activeMeshVersion,
                mode: mode
            )
        }
    }
}

private final class ProgressListener: NativeRuntimeProgressListener, @unchecked Sendable {
    private let onProgress: @Sendable (NativeRuntimeDownloadProgress) -> Void

    init(_ onProgress: @escaping @Sendable (NativeRuntimeDownloadProgress) -> Void) {
        self.onProgress = onProgress
    }

    func onProgress(event: NativeRuntimeDownloadProgressNative) {
        onProgress(event)
    }
}

private func runBlocking<T>(_ work: @escaping () throws -> T) async throws -> T {
    try await withCheckedThrowingContinuation { continuation in
        DispatchQueue.global().async(flags: .inheritQoS) {
            do {
                continuation.resume(returning: try work())
            } catch {
                continuation.resume(throwing: error)
            }
        }
    }
}
#else
#error("MeshLLM Swift SDK requires MeshLLMFFI.xcframework. Build it with sdk/swift/scripts/build-xcframework.sh before building, testing, or integrating the SDK.")
#endif

File: sdk/kotlin/src/main/kotlin/ai/meshllm/NativeRuntime.kt
package ai.meshllm

import java.io.File
import uniffi.mesh_ffi.InstalledNativeRuntimeNative
import uniffi.mesh_ffi.NativeRuntimeDownloadProgressNative
import uniffi.mesh_ffi.NativeRuntimeInstallOptionsNative
import uniffi.mesh_ffi.NativeRuntimeInstallOutcomeNative
import uniffi.mesh_ffi.NativeRuntimeProgressListener
import uniffi.mesh_ffi.NativeRuntimePruneModeNative
import uniffi.mesh_ffi.NativeRuntimePruneResultNative
import uniffi.mesh_ffi.NativeRuntimeVerificationPolicyNative
import uniffi.mesh_ffi.currentMeshVersion
import uniffi.mesh_ffi.currentSkippyAbiVersion
import uniffi.mesh_ffi.installNativeRuntime
import uniffi.mesh_ffi.installedNativeRuntimes
import uniffi.mesh_ffi.pruneNativeRuntimes
import uniffi.mesh_ffi.removeNativeRuntime

typealias NativeRuntimeInstallOptions = NativeRuntimeInstallOptionsNative
typealias NativeRuntimeInstallOutcome = NativeRuntimeInstallOutcomeNative
typealias NativeRuntimeDownloadProgress = NativeRuntimeDownloadProgressNative
typealias InstalledNativeRuntime = InstalledNativeRuntimeNative
typealias NativeRuntimePruneMode = NativeRuntimePruneModeNative
typealias NativeRuntimePruneResult = NativeRuntimePruneResultNative
typealias NativeRuntimeVerificationPolicy = NativeRuntimeVerificationPolicyNative

data class NativeRuntimeResolveOptions(
    val artifactDir: File? = null,
    val searchDirs: List<File> = emptyList(),
    val cacheDir: File? = null,
    val allowDownload: Boolean = false,
    val selection: String = "recommended",
)

object NativeRuntime {
    val meshVersion: String
        get() = currentMeshVersion()

    val skippyAbiVersion: String
        get() = currentSkippyAbiVersion()

    fun install(
        options: NativeRuntimeInstallOptions = defaultInstallOptions(),
        onProgress: ((NativeRuntimeDownloadProgress) -> Unit)? = null,
    ): NativeRuntimeInstallOutcome =
        installNativeRuntime(options, onProgress?.let(::ProgressListener))

    fun resolve(
        options: NativeRuntimeResolveOptions = NativeRuntimeResolveOptions(),
        onProgress: ((NativeRuntimeDownloadProgress) -> Unit)? = null,
    ): InstalledNativeRuntime {
        val bundleDirs = buildList {
            options.artifactDir?.let { add(it.path) }
            addAll(options.searchDirs.map(File::getPath))
        }
        return install(
            defaultInstallOptions(
                selection = options.selection,
                bundleDirs = bundleDirs,
                cacheDir = options.cacheDir?.path,
                allowDownload = options.allowDownload,
            ),
            onProgress,
        ).runtime
    }

    fun validate(artifactDir: File, cacheDir: File? = null): InstalledNativeRuntime =
        resolve(
            NativeRuntimeResolveOptions(
                artifactDir = artifactDir,
                cacheDir = cacheDir,
                allowDownload = false,
            )
        )

    fun installed(cacheDir: File? = null): List<InstalledNativeRuntime> =
        installedNativeRuntimes(cacheDir?.path)

    fun remove(
        meshVersion: String,
        nativeRuntimeId: String,
        cacheDir: File? = null,
    ): Boolean =
        removeNativeRuntime(cacheDir?.path, meshVersion, nativeRuntimeId)

    fun prune(
        cacheDir: File? = null,
        activeMeshVersion: String? = null,
        mode: NativeRuntimePruneMode = NativeRuntimePruneMode.KEEP_ACTIVE_AND_PREVIOUS,
    ): NativeRuntimePruneResult =
        pruneNativeRuntimes(cacheDir?.path, activeMeshVersion, mode)

    fun defaultInstallOptions(
        meshVersion: String? = null,
        skippyAbiVersion: String? = null,
        selection: String = "recommended",
        manifestPath: String? = null,
        manifestUrl: String? = null,
        bundleDirs: List<String> = emptyList(),
        cacheDir: String? = null,
        verificationPolicy: NativeRuntimeVerificationPolicy = NativeRuntimeVerificationPolicy.REQUIRE_CHECKSUM,
        allowDownload: Boolean = true,
    ): NativeRuntimeInstallOptions =
        NativeRuntimeInstallOptions(
            meshVersion = meshVersion,
            skippyAbiVersion = skippyAbiVersion,
            selection = selection,
            manifestPath = manifestPath,
            manifestUrl = manifestUrl,
            bundleDirs = bundleDirs,
            cacheDir = cacheDir,
            verificationPolicy = verificationPolicy,
            allowDownload = allowDownload,
        )
}

private class ProgressListener(
    private val onProgress: (NativeRuntimeDownloadProgress) -> Unit,
) : NativeRuntimeProgressListener {
    override fun onProgress(event: NativeRuntimeDownloadProgressNative) {
        onProgress.invoke(event)
    }
}
